core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp

```diff
@@ -241,31 +265,46 @@ struct DeduceCudaLaunchMechanism {
       (sizeof(DriverType) < CudaTraits::KernelArgumentLimit
            ? CudaLaunchMechanism::LocalMemory
            : CudaLaunchMechanism::Default) |
+#ifndef KOKKOS_IMPL_USE_GRID_CONSTANT
       (sizeof(DriverType) < CudaTraits::ConstantMemoryUsage
            ? CudaLaunchMechanism::ConstantMemory
            : CudaLaunchMechanism::Default) |
+#endif
       CudaLaunchMechanism::GlobalMemory;
 
   static constexpr CudaLaunchMechanism requested_launch_mechanism =
+#ifdef KOKKOS_IMPL_USE_GRID_CONSTANT
+      CudaLaunchMechanism::LocalMemory |
+#else
       (((property & light_weight) == light_weight)
            ? CudaLaunchMechanism::LocalMemory
            : CudaLaunchMechanism::ConstantMemory) |
+#endif
       CudaLaunchMechanism::GlobalMemory;
 
   static constexpr CudaLaunchMechanism default_launch_mechanism =
+#ifdef KOKKOS_IMPL_USE_GRID_CONSTANT
+      (sizeof(DriverType) < CudaTraits::KernelArgumentLimit)
+          ? CudaLaunchMechanism::LocalMemory
+          : CudaLaunchMechanism::GlobalMemory;
+#else
       // BuildValidMask
       (sizeof(DriverType) < CudaTraits::ConstantMemoryUseThreshold)
           ? CudaLaunchMechanism::LocalMemory
           : ((sizeof(DriverType) < CudaTraits::ConstantMemoryUsage)
                  ? CudaLaunchMechanism::ConstantMemory
                  : CudaLaunchMechanism::GlobalMemory);
+#endif
 
   //              None                LightWeight    HeavyWeight
   // F<UseT       LCG LCG L  L        LCG  LG L  L    LCG  CG L  C
   // UseT<F<KAL   LCG LCG C  C        LCG  LG C  L    LCG  CG C  C
   // Kal<F<CMU     CG LCG C  C         CG  LG C  G     CG  CG C  C
   // CMU<F          G LCG G  G          G  LG G  G      G  CG G  G
   static constexpr CudaLaunchMechanism launch_mechanism =
+#ifdef KOKKOS_IMPL_USE_GRID_CONSTANT
+      default_launch_mechanism;
```

Comment:
You should include the comment above in the #ifndef branch
```C++
#ifdef KOKKOS_IMPL_USE_GRID_CONSTANT
static constexpr CudaLaunchMechanism launch_mechanism = default_launch_mechanism;
#else
<the old code>
#endif

Reply:
Also this means we would always ignore the `{Light,Heavy}Weight property.
Please make sure you discuss it in the description (why we don't just use that hint as long as it is valid for instance) and add a comment here in the code saying we ignore it.